src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -343,64 +345,35 @@ generateX x
              $ hoistEither $ errorNoSuggestions (ErrorFunctionWrongArgs a x fErr [])
            annotate Map.empty resT $ \a' -> Prim a' p
 
-    -- Cases require all their alternatives to have the same type, as well as the scrutinee and the alternatives to have
-    -- the same temporality.
-    -- This may be more restrictive than necessary, since lets allow different temporalities, but hopefully will not be an issue.
-    --
-    -- TODO: cases with nested constructors must be converted to multiple nested cases.
-    -- For example:
-    -- > case x
-    -- >  | Some (Some v, None) -> v
-    -- >  | Some (None, Some v) -> v
-    -- >  | _                   -> 0
-    -- > end
-    --
-    -- this should be flattened to
-    -- > case x
-    -- > | Some y
-    -- > -> case y
-    -- >    | (l,r)
-    -- >    -> case l
-    -- >       | Some v
-    -- >       -> case r
-    -- >          | Some _
-    -- >          -> 0
-    -- >          | None
-    -- >          -> v
-    -- >          end
-    -- >       | None
-    -- >       -> case r
-    -- >          | Some v
-    -- >          -> v
-    -- >          | None
-    -- >          -> 0
-    -- >          end
-    -- >       end
-    -- >    end
-    -- > | None
-    -- > -> 0
+    -- Cases require:
+    --  * alternatives to have "join-able" types, i.e. no mixing of aggregates and elements.
     --
     Case a scrut pats
      -> do (scrut', sub) <- generateX scrut
            (pats', subs) <- unzip <$> mapM (generateP $ annResult $ annotOfExp scrut') pats
 
-           resT <- case pats' of
-                    [] -> hoistEither
-                        $ errorNoSuggestions
-                        $ ErrorEmptyCase a x
-                    ((_,alt):_)
-                       -> return $ annResult $ annotOfExp alt
+           let scrutT  =  annResult $ annotOfExp scrut'
+           resT        <- case pats' of
+                           [] -> hoistEither
+                               $ errorNoSuggestions
+                               $ ErrorEmptyCase a x
+                           ((_,alt):_)
+                              -> return $ annResult $ annotOfExp alt
```

Comment:
hmm - if the first alternative is Pure and the others are Element, the return should be Element.
I think we need a fresh type variable for the return, then go through and say the return = the join of all the scruts and alts together.
actually that's rather annoying, because you need a fresh variable for each scrut and alt..

```
case S
 | 1 -> A1
 | 2 -> A2

==>
Constraints:
 T1 = Join S1 A1
 T2 = Join T1 A2

Return temporality = T2

```